gcloud/datastore/helpers.py

```diff
@@ -155,6 +155,48 @@ def entity_from_protobuf(pb):
     return entity
 
 
+def _set_pb_meaning_from_entity(entity, name, value, value_pb,
+                                is_list=False):
+    """Add meaning information (from an entity) to a protobuf.
+
+    :type entity: :class:`gcloud.datastore.entity.Entity`
+    :param entity: The entity to be turned into a protobuf.
+
+    :type name: string
+    :param name: The name of the property.
+
+    :type value: object
+    :param value: The current value stored as property ``name``.
+
+    :type value_pb: :class:`gcloud.datastore._generated.entity_pb2.Value`
+    :param value_pb: The protobuf value to add meaning / meanings to.
+
+    :type is_list: bool
+    :param is_list: (Optional) Boolean indicating if the ``value`` is
+                    a list value.
+    """
+    if name not in entity._meanings:
+        return
+
+    meaning, orig_value = entity._meanings[name]
+    # Only add the meaning back to the protobuf if the value is
+    # unchanged from when it was originally read from the API.
+    if orig_value is not value:
+        return
+
+    # For lists, we set meaning on each sub-element.
+    if is_list:
```

Comment:
IIUC, if we see a value A, we'll only preserve the meaning if the value was set to A before. 

However, if we had [A, B] before, then we got [A, C], we'll drop the meaning for both A and C. However, couldn't we detect that A is still the same, and preserve the meaning? Basically, don't return above in the event of a list and instead try to preserve things?

Note: I think this is already much nicer than before, so I don't think this is necessary until users start to hit it. Just wanted to point it out as somewhere we could maybe make things a little nicer.